### ISSUE TITLE
Inheriting from Grape::API yields anonymous class

### ISSUE THREAD

**Issue by delner** [ruby-grape/grape#1825]
With the release of version 1.2.0, the following class definition yields an anonymous class:

```
> class Foo < ::Grape::API; end
=> nil
> Foo
=> #<Class:0x0000560d0a089978>
```

This breaks some code that depends on referencing `Foo` as a constant to return a well-known type, not an anonymous class. A common case would be calling `Foo.name` which previously returned `'Foo'` would now return `nil`.

A workaround for this, [as alluded to by the release notes](https://github.com/ruby-grape/grape/blob/3cf89cf1f253a36c73d4b9e8a6b2aca673f0b915/UPGRADING.md), is to convert all such instances of `::Grape::API` to `::Grape::API::Instance` instead. However this doesn't really address the issue in cases where inheriting from `Grape::API` is still considered acceptable practice.

**Comment by dblock** [comment]
Good catch, I am not sure what the best way to fix it is, but let's make sure to make some specs, cc: @myxoh.

**Comment by myxoh** [comment]
Could you tell me which version of Ruby you are using?

On MRI Ruby 2.3.3

![image](https://user-images.githubusercontent.com/5561861/49079687-2f92af00-f239-11e8-9692-e48874b94506.png)

On MRI Ruby 2.5.1

![image](https://user-images.githubusercontent.com/5561861/49079737-55b84f00-f239-11e8-8ae6-2db07d728290.png)

So

**Comment by delner** [comment]
I was having trouble reliably replicating this the same way you did (irb), using Ruby 2.3.7. However, this "anonymous class with no name" problem can be replicated with the following Minitest, when working with the API class from an ActiveSupport notification:

```
require 'minitest'
require 'minitest/autorun'
require 'rack/test'
require 'grape'

class TestGrapeAPI < Grape::API
  namespace :base do
    desc 'Returns a success message'
    get :success do
      'OK'
    end
  end
end

class GrapeAPITest < MiniTest::Test
  include Rack::Test::Methods

  def setup
    ::ActiveSupport::Notifications.subscribe('endpoint_run.grape', &method(:callback))
  end

  def app
    TestGrapeAPI
  end

  def test_api_name
    get '/base/success'
    assert last_response.ok?
    assert_equal('OK', last_response.body)
  end

  def callback(name, start, finish, id, payload)
    api = payload[:endpoint].options[:for]
    # Expect the API name to not be nil
    refute_nil(api.name)
  end
end
```

![screenshot from 2018-11-27 11-25-06](https://user-images.githubusercontent.com/3237131/49095950-8303fd80-f237-11e8-90d9-fed80810be59.png)

Hopefully this helps assess the problem.

**Comment by myxoh** [comment]
Ok, yes. The problem you are talking about is that the MOUNTED app is anonymous. Which is intentional. as we are now mounting instances, rather than the class itself. 

So: I shouldn't make the anonymous class respond to `name` with the name of the original class, because that'd generate all sorts of trouble, they are two different classes. 

Without me changing anything, in your test you can do:

`api = payload[:endpoint].options[:for].base`

to get a reference to the original class, if you do that your test will pass.

If this is not an option for you (calling base), I could make it so that the instances are constantized, for the time being I feel that it is more trouble than it's worth.

Is `mounted_instance.base` good enough for your needs?

If so, I can provide better documentation for it

**Comment by myxoh** [comment]
(Related to the above ^) 
The following code passes:

```
require 'minitest'
require 'minitest/autorun'
require 'rack/test'
require 'grape'

class TestGrapeAPI < Grape::API
  namespace :base do
    desc 'Returns a success message'
    get :success do
      'OK'
    end
  end
end

class GrapeAPITest < MiniTest::Test
  include Rack::Test::Methods

  def setup
    ::ActiveSupport::Notifications.subscribe('endpoint_run.grape', &method(:callback))
  end

  def app
    TestGrapeAPI
  end

  def test_api_name
    get '/base/success'
    assert last_response.ok?
    assert_equal('OK', last_response.body)
  end

  def callback(name, start, finish, id, payload)
    api = payload[:endpoint].options[:for].base # difference is this line calls `.base`
    # Expect the API name to not be nil
    refute_nil(api.name, "#name for #{api} is nil!")
  end
end
```

**Comment by dblock** [comment]
Should this anonymous class respond to `name` and provide something like `API Name#instance ID`?

**Comment by myxoh** [comment]
We could try, my concern is that is still an instance of the class `Class` where the keywoard `name` has a well defined meaning. Not sure what would be the side effects of it incorrectly identifying as `API Name`

I would be happy to provide alternative methods like `api_name`, and potentially giving the instance a name like you suggested (`ApiConstant::Instance_1`). Alternative, I guess, we could try defining `name` and seeing what happens.

**Comment by dblock** [comment]
Your call @myxoh if you want to try something. We could at least document in UPGRADING what one should do if they were using `Grape::API#name`?

**Comment by dblock** [comment]
@myxoh Check out the diff in https://github.com/DataDog/dd-trace-rb/pull/639/files - do you think that's the right way to do it?

**Comment by delner** [comment]
Previously we had actually did `to_s` instead of `name`. Maybe overriding the behavior of `to_s` is less invasive that overriding the class `name` function?

Also, using `base` seemed to work fine, so there is a way of working around the name issue I was having. (See the linked PR above.)

**Comment by myxoh** [comment]
I agree with overriding `to_s` in principle. 
I'll create a PR in a bit that does this.

Regarding your diff @dblock , I think that's sensible. I wish I could think of something that could do without the if. Perhaps `to_s` if we override that works for both versions?

Finally, I'll create a note on the upgrading to reflect this

**Comment by delner** [comment]
A `to_s` override that would allow 1.1 and 1.2 to work with the same code would be awesome. Would like to avoid the `if` too.

**Comment by myxoh** [comment]
Apologize for the delay. 
If you could please take a quick look at this PR: https://github.com/ruby-grape/grape/pull/1826

**Comment by JaniJegoroff** [comment]
@delner Thanks, there was something fishy going on with my tests after updating to `1.2.0` and just confirmed it's related to this issue.

**Comment by myxoh** [comment]
@dblock As soon as I can I'll release it

**Comment by myxoh** [comment]
Released

**Comment by delner** [comment]
Thanks @myxoh and @dblock for the quick turnaround on this!

**Comment by JaniJegoroff** [comment]
Still some of my tests are failing after updating to `1.2.1`. When I make this change everything works again
```
# not OK
module MyMockServer
  class API < Grape::API

# OK
module MyMockServer
  class API < Grape::API::Instance
```

**Comment by myxoh** [comment]
Could you please describe your issue in a little more detail?
I'm not sure what error you are seeing

**Comment by JaniJegoroff** [comment]
@myxoh I wish I could. I'm testing web application against local mock server that I implemented with `grape`. Unfortunately, I don't see errors from grape directly so it's not easy to say what's happening deep down. I have used several versions in the past before `1.2.0` successfully and I can see some tests failing in versions `1.2.0` and `1.2.1` only (tests are passing again if I modify code as mentioned earlier).

I'm trying to find more time to debug and create an example project to reproduce the issue.

**Comment by myxoh** [comment]
@JaniJegoroff I'm not sure I can fix it without an example :cry: . When you do, could you please open a separate issue?

**Comment by JaniJegoroff** [comment]
@myxoh I fully understand and will create separate issue in due time.

**Comment by JaniJegoroff** [comment]
@myxoh Looks like release 1.2.2 resolves issues I had with 1.2.0/1.2.1 👍
